File: Final/Code/Project_Interger_Partition/Bai_3/A/a.py
def is_self_conjugate(partition):
    for i in range(len(partition)):
        if partition[i] < i + 1:
            return False
    return True

def generate_partitions(n, max_val, current, k, result):
    if n == 0 and len(current) == k and is_self_conjugate(current):
        result.append(current[:])
        return
    for i in range(min(n, max_val), 0, -1):
        current.append(i)
        generate_partitions(n - i, i, current, k, result)
        current.pop()

n = int(input("Nhập n: "))
k = int(input("Nhập k: "))
result = []
generate_partitions(n, n, [], k, result)
for p in result:
    print(p)
print(f"Tổng số phân hoạch tự liên hợp có {k} phần: {len(result)}")

File: Final/Code/Project_Interger_Partition/Bai_3/A/a.cpp
#include <iostream>
#include <vector>

using namespace std;

// Kiểm tra xem phân hoạch có phải tự liên hợp không
bool isSelfConjugate(const vector<int>& part) {
    int n = part.size();
    for (int i = 0; i < n; ++i) {
        if (part[i] < i + 1) return false;
    }
    return true;
}

void generatePartitions(int n, int max, vector<int>& curr, int k, int& count) {
    if (n == 0 && curr.size() == k && isSelfConjugate(curr)) {
        ++count;
        for (int v : curr) cout << v << " ";
        cout << endl;
        return;
    }
    for (int i = min(n, max); i >= 1; --i) {
        curr.push_back(i);
        generatePartitions(n - i, i, curr, k, count);
        curr.pop_back();
    }
}

int main() {
    int n;
    int k;
    cout << "Nhập n: ";
    cin >> n;
    cout << "Nhập k: ";
    cin >> k;
    vector<int> curr;
    int count = 0;
    generatePartitions(n, n, curr, k, count);
    cout << "Tổng số phân hoạch tự liên hợp có " << k << " phần: " << count << endl;
    return 0;
}
